kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,84 @@
+package kyo
+
+import Layers.internal.*
+import kyo.Flat.unsafe
+import kyo.core.*
+import scala.annotation.targetName
+
+sealed trait Layer[-In, +Out, -S]:
+    self =>
+
+    def >>>[In2, Out2, S2](that: Layer[Out & In2, Out2, S2]): Layer[In & In2, Out2, S & S2]       = To(self, that)
+    def ++[In2, Out2, S2](that: Layer[In2, Out2, S2]): Layer[In & In2, Out & Out2, S & S2]        = And(self, that)
+    def >+>[In2, Out2, S2](that: Layer[Out & In2, Out2, S2]): Layer[In & In2, Out & Out2, S & S2] = self ++ (self >>> that)
+
+    private[kyo] def doRun(
+        memoMap: scala.collection.mutable.Map[Layer[?, ?, ?], Any] = scala.collection.mutable.Map.empty[Layer[?, ?, ?], Any]
+    ): TypeMap[Out] < (S & Envs[In]) =
+        type Expected = TypeMap[Out] < (S & Envs[In])
+        memoMap.get(self) match
+            case Some(result) => result.asInstanceOf[Expected]
+            case None =>
+                self match
+                    case And(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- rhs.doRun(memoMap)
+                            yield leftResult.union(rightResult)
+                        }.asInstanceOf[Expected]
+
+                    case To(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- Envs.runTypeMap(leftResult)(rhs.doRun(memoMap))(using unsafe.bypass, Envs.bypass)
```

Comment:
Do we need the `Flat` bypass here now that `TypeMap` has an evidence for it?

Reply:
Nope, I can use `summon` instead.